lib/gc/ExecutionEngine/CPURuntime/Microkernel/BrgemmOnednn.cpp

```diff
@@ -0,0 +1,163 @@
+//===-- BrgemmOnednn.cpp - BRGEMM Onednn Implementation ---------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <assert.h>
+#include <math.h>
+#include <memory>
+#include <shared_mutex>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <vector>
+#ifdef _WIN32
+#include <windows.h>
+#else
+#include <unistd.h>
+#endif
+
+// manually include xbyak header here to avoid no-exception compile issue
+#define XBYAK_NO_EXCEPTION
+#include <cpu/x64/xbyak/xbyak.h> // NOLINT
+#undef XBYAK_NO_EXCEPTION
+
+#include <cpu/x64/amx_tile_configure.hpp>
+#include <cpu/x64/brgemm/brgemm.hpp>
+#include <cpu/x64/brgemm/brgemm_types.hpp>
+#include <cpu/x64/cpu_isa_traits.hpp>
+
+#include "gc/ExecutionEngine/CPURuntime/Microkernel/BrgemmInterface.h"
+
+using namespace dnnl::impl::cpu::x64;
+
+namespace dnnl {
+namespace impl {
+namespace graph {
+namespace utils {
+// dummy definition for DNNL lite linkage
+__attribute__((weak)) void print_verbose_header() {}
+} // namespace utils
+} // namespace graph
+} // namespace impl
+} // namespace dnnl
+
+static constexpr int PALETTE_SIZE = 64;
+
+using read_lock_guard_t = std::shared_lock<std::shared_mutex>;
+using write_lock_guard_t = std::unique_lock<std::shared_mutex>;
+static std::shared_mutex g_brgemm_lock;
+
+static std::vector<brgemm_desc_t> g_brgemm_desc_list;
+static std::vector<brgemm_kernel_t *> g_brgemm_kernel_list;
+static std::vector<std::unique_ptr<char[]>> g_brgemm_palette;
+
+// TODO(haixin): use syscall to determine page size?
+static constexpr size_t SCRATCH_SIZE = 2 * 4096;
+// TODO(haixin): need to use custom thread management for scratch in the future?
+static thread_local char scratch[SCRATCH_SIZE] = {0};
+
+extern "C" {
+
+int64_t dnnl_brgemm_dispatch(int64_t M, int64_t N, int64_t K, int64_t LDA,
+                             int64_t LDB, int64_t LDC, int64_t stride_a,
+                             int64_t stride_b, float beta, int64_t dtypeA,
+                             int64_t dtypeB) {
+  brgemm_desc_t desc;
+  brgemm_kernel_t *kernel;
+
+  auto dnnl_dtypeA = static_cast<dnnl_data_type_t>(dtypeA);
+  auto dnnl_dtypeB = static_cast<dnnl_data_type_t>(dtypeB);
+  int64_t dtypeA_size = dnnl::impl::types::data_type_size(dnnl_dtypeA);
+  int64_t dtypeB_size = dnnl::impl::types::data_type_size(dnnl_dtypeB);
+  brgemm_strides_t stride_info{stride_a * dtypeA_size, stride_b * dtypeB_size};
+
+  dnnl::impl::status_t status = brgemm_desc_init(
```

Comment:
How to reuse the same `desc` or kernel instance, the current implementation will always create a new one and push to the global storage.

Reply:
Suggest to rename `little` to `brgemm` which is more meaningful.

---------------------
The name is `lite` which is borrowed from GC V1, and we might add more stuff to it other than BRGEMM in the future. So maybe we could keep the name?

How to reuse the same desc or kernel instance, the current implementation will always create a new one and push to the global storage.

-----------------
Currently kernel reuse is done in MLIR level. In opt pass `EarlyDispatchMicrokernel` (not merged yet), it would try to reuse kernels with identical input parameters and would only invoke `dispatch` once for them.